test/init_test.go

```diff
@@ -0,0 +1,183 @@
+
+import (
+	"flag"
+	"fmt"
+	"strings"
+	"sync"
+	"testing"
+
+	knativetest "github.com/knative/pkg/test"
+	"github.com/knative/pkg/test/logging"
+	"github.com/tektoncd/pipeline/pkg/names"
+	"golang.org/x/xerrors"
+	yaml "gopkg.in/yaml.v2"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+)
+
+var initMetrics sync.Once
+
+func setup(t *testing.T) (*clients, string) {
+	t.Helper()
+	namespace := names.SimpleNameGenerator.RestrictLengthWithRandomSuffix("arrakis")
+
+	initializeLogsAndMetrics(t)
+
+	c := newClients(t, knativetest.Flags.Kubeconfig, knativetest.Flags.Cluster)
+	createNamespace(t, namespace, c.KubeClient)
+	verifyDefaultServiceAccountExists(t, namespace, c.KubeClient)
+	return c, namespace
+}
+
+func tearDown(t *testing.T, cs *clients, namespace string) {
+	t.Helper()
+	if cs.KubeClient == nil {
+		return
+	}
+	if t.Failed() {
+		header(t.Logf, fmt.Sprintf("Dumping objects from %s", namespace))
+		bs, err := getCRDYaml(cs, namespace)
+		if err != nil {
+			t.Error(err)
+		} else {
+			t.Log(string(bs))
+		}
+	}
+
+	t.Logf("Deleting namespace %s", namespace)
+	if err := cs.KubeClient.CoreV1().Namespaces().Delete(namespace, &metav1.DeleteOptions{}); err != nil {
+		t.Errorf("Failed to delete namespace %s: %s", namespace, err)
+	}
+}
+
+func header(logf logging.FormatLogger, text string) {
+	left := "### "
+	right := " ###"
+	txt := left + text + right
+	bar := strings.Repeat("#", len(txt))
+	logf(bar)
+	logf(txt)
+	logf(bar)
+}
+
+func initializeLogsAndMetrics(t *testing.T) {
+	t.Helper()
+	initMetrics.Do(func() {
+		flag.Parse()
+		if err := flag.Set("alsologtostderr", "true"); err != nil {
+			t.Fatalf("Failed to set 'alsologtostderr' flag to 'true': %s", err)
+		}
+		logging.InitializeLogger(knativetest.Flags.LogVerbose)
+
+		if knativetest.Flags.EmitMetrics {
+			logging.InitializeMetricExporter(t.Name())
+		}
+	})
+}
+
+func createNamespace(t *testing.T, namespace string, kubeClient kubernetes.Interface) {
+	t.Helper()
+	t.Logf("Create namespace %s to deploy to", namespace)
+	if _, err := kubeClient.CoreV1().Namespaces().Create(&corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: namespace,
+		},
+	}); err != nil {
+		t.Fatalf("Failed to create namespace %s for tests: %s", namespace, err)
+	}
+}
+
+func verifyDefaultServiceAccountExists(t *testing.T, namespace string, kubeClient kubernetes.Interface) {
+	t.Helper()
+	defaultSA := "default"
+	t.Logf("Verify SA %s is created in namespace %s", defaultSA, namespace)
+
+	if err := wait.PollImmediate(interval, timeout, func() (bool, error) {
+		_, err := kubeClient.CoreV1().ServiceAccounts(namespace).Get(defaultSA, metav1.GetOptions{})
+		if err != nil && errors.IsNotFound(err) {
+			return false, nil
+		}
+		return true, err
+	}); err != nil {
+		t.Fatalf("Failed to get SA %q in namespace %q for tests: %s", defaultSA, namespace, err)
+	}
```

Comment:
hm, can you explain a bit more about the purpose of this check? i would think it would be safe to assume the default account always exists?

Reply:
I put it in because that's what the [Pipelines project](https://github.com/tektoncd/pipeline/blob/master/test/init_test.go#L52) does 😅 But I can definitely remove this check if it isn't necessary.

Reply:
Maybe pipelines needs this in case no service account is provided so that there is a fall back?